src/scanner/utility.ts

```diff
@@ -0,0 +1,1031 @@
+let blockTime;
+
+// Convert a block number to date
+async function convertBlockToDate(api, blockNumber, date) {
+	blockTime = blockTime ? blockTime : await api.consts.babe.expectedBlockTime;
+	const value = blockTime.mul(bnToBn(blockNumber)).toNumber();
+	const time = extractTime(Math.abs(value));
+	const { days } = time;
+	const d = new Date(date);
+	d.setDate(d.getDate() + days);
+	return d;
+}
+
+// store in db all the relevant data for mint unique nft
+export async function trackUniqueMintData(
+	eventData,
+	api,
+	params,
+	date,
+	owner,
+	txHash,
+	blockNumber
+) {
+	try {
+		const tokenId = JSON.stringify(eventData[1]); // tokenId in format [17,5,0] - [collectionId, seriesId, serialNo]
+		const imgUrl = api.registry
+			.createType(params[3].type, params[3].value)
+			.toHuman();
+		const tokenData = {
+			eventData: {
+				imgUrl: imgUrl,
+				date: date,
+				owner: owner,
+				txHash: txHash,
+			},
+			eventType: "NFT_CREATED",
+		};
+		const type = 0;
+		await trackEventData(
+			tokenId,
+			type,
+			blockNumber,
+			JSON.stringify(tokenData),
+			owner
+		);
+	} catch (e) {
+		console.log(
+			`Error tracking unique mint data with params ${JSON.stringify(
+				params
+			)}, error ${e}`
+		);
+	}
+}
+
+// mint series add all the tokens as per the count argument in the db
+export async function trackTokenSeriesData(
+	eventData,
+	api,
+	params,
+	date,
+	owner,
+	txHash,
+	blockNumber
+) {
+	try {
+		const collectionId = eventData[0];
+		const seriesId = eventData[1];
+		const noOfTokens = eventData[2]; // quantity
+		const imgUrl = api.registry
+			.createType(params[4].type, params[4].value)
+			.toHuman();
+		const tokenData = {
+			eventData: {
+				imgUrl: imgUrl,
+				date: date,
+				owner: owner,
+				txHash: txHash,
+			},
+			eventType: "NFT_CREATED",
+		};
+		const type = 0; // nft token data
+		await extractTokenList(
+			0,
+			noOfTokens,
+			collectionId,
+			seriesId,
+			type,
+			blockNumber,
+			tokenData,
+			owner
+		);
+	} catch (e) {
+		console.log(
+			`Error tracking token series data with params ${JSON.stringify(
+				params
+			)}, error ${e}`
+		);
+	}
+}
+
+// mint series add additional nft for a given collection, find next serialNumber at this blockhash and add more
+export async function trackAdditionalTokenData(
+	params,
+	eventData,
+	api,
+	blockHash,
+	date,
+	owner,
+	txHash,
+	blockNumber
+) {
+	try {
+		const collectionId = eventData[0];
+		const seriesId = eventData[1];
+		const noOfTokens = eventData[2]; // quantity
+		const nextSerialNumber = (
+			await api.query.nft.nextSerialNumber.at(blockHash, collectionId, seriesId)
+		).toNumber();
+		const imgUrl = api.query.nft.seriesMetadataScheme
+			? (
+					await api.query.nft.seriesMetadataScheme.at(
+						blockHash,
+						collectionId,
+						seriesId
+					)
+			  ).toHuman()
+			: (
+					await api.query.nft.seriesMetadataURI.at(
+						blockHash,
+						collectionId,
+						seriesId
+					)
+			  ).toHuman();
+		const tokenData = {
```

Comment:
instead of await one after another, if they can happen in parallel then you can utilise the `Promise.all`, that will shave off a few `ms`

```
const [nextSerialNumber, imgUrl] = await Promise.all([
await api.query.nft.nextSerialNumber.at(blockHash, collectionId, seriesId),
api.query.nft.seriesMetadataScheme
			? (
					await api.query.nft.seriesMetadataScheme.at(
						blockHash,
						collectionId,
						seriesId
					)
			  ).toHuman()
			: (
					await api.query.nft.seriesMetadataURI.at(
						blockHash,
						collectionId,
						seriesId
					)
			  ).toHuman()
])
```

I might have not picked up everything, so please use this pattern anywhere if it is applicable. Will help speed up script

Reply:
sure..